### PR TITLE
feat: modernaization radio widget

### LINKED ISSUE #817: Modernize Radio widget
**Is your feature request related to a problem? Please describe.**
Modernize radio widget per [docs](https://docs.ensembleui.com/widgets/radio-group) 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by sagardspeed2**
## Describe your changes
Added item-template support in Radio widget

## Issue ticket number and link

Closes #817 

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added tests
- [ ] I have added a changeset `pnpm changeset add`
- [ ] I have added example usage in the kitchen sink app

### CODE REVIEW

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 1]
```diff
@@ -21,15 +33,42 @@ export type RadioWidgetProps = FormInputProps<string> &
...
+  // update form field initial value
+  const formInstance = Form.useFormInstance();
+
+  useEffect(() => {
+    if (formInstance) {
+      formInstance.setFieldsValue({
+        [values?.id ?? values?.label ?? ""]: value,
+      });
+    }
+  }, [value, formInstance]);
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:66]: Why is this needed?

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 2]
```diff
@@ -38,19 +77,58 @@ export const RadioWidget: React.FC<RadioWidgetProps> = (props) => {
...
+          style: getStyle(item.styles as object),
+        });
       });
     }
-  }, [value, formInstance]);
+
+    if (isObject(itemTemplate) && !isEmpty(namedData)) {
+      namedData.forEach((item: unknown) => {
+        const typedItem = get(item, itemTemplate.name) as CustomScope;
+        const evaluateValue = evaluate<string | number>(
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:109]: Do not use.

**Review by evshi — CHANGES_REQUESTED** [review]

File: packages/runtime/src/widgets/Form/Radio.tsx (hunk 1 above)
Comment (sagardspeed2) [reply packages/runtime/src/widgets/Form/Radio.tsx:66]: @evshi this is needed to set the initial values of radio widget in the form instance, so it will validate that values or we can get those values on form submit time.

we did this in all form elements like, dropdown, radio, checkbox, textInput, Date and so on

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 3]
```diff
@@ -21,15 +31,42 @@ export type RadioWidgetProps = FormInputProps<string> &
...
-    props.id,
+    { ...rest, initialValue: rest.value, value, widgetName },
+    rest.id,
     { setValue },
   );
-  const action = useEnsembleAction(props.onChange);
 
+  // update initial value
+  useEffect(() => {
+    setValue(values?.initialValue?.toString());
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:52]: Value may not be a string so don't force it to one.

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 4]
```diff
@@ -38,19 +75,60 @@ export const RadioWidget: React.FC<RadioWidgetProps> = (props) => {
...
+    if (isObject(itemTemplate) && !isEmpty(namedData)) {
+      namedData.forEach((item: unknown) => {
+        const typedItem = get(item, itemTemplate.name) as CustomScope;
+        const evaluateValue = get(
+          item,
+          itemTemplate.value
+            .toString()
+            // eslint-disable-next-line prefer-named-capture-group
+            .replace(/['"]?\$\{([^}]*)\}['"]?/g, "$1"),
+        ) as string | number;
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:113]: Again this is forcing value as a string?

**Review by evshi — CHANGES_REQUESTED** [review]
Please write some tests that include values of different types - string, number, object.

File: packages/runtime/src/widgets/Form/Radio.tsx (hunk 4 above)
Comment (sagardspeed2) [reply packages/runtime/src/widgets/Form/Radio.tsx:113]: @evshi here, `itemTemplate.value` will be always string, for example,

```yaml
item-template:
    data: getData.body.users.slice(0, 4)
    name: user
    value: ${user.email}
    template:
      Row:
        styles:
          gap: 10px
          paddingBlock: 10px
          alignItems: center
        children:
          - Image:
              width: 25
              height: 25
              source: ${user.image}
          - Text:
              text: ${user.firstName} ${user.lastName}
```

where user.email will be always string and indicate that which value need to assign that iterate, here it says that use `user.email` as a value.

Comment (evshi) [reply packages/runtime/src/widgets/Form/Radio.tsx:113]: Oh I misread this. You shouldn't need to find the evaluated value - `useItemTemplate` does that for you in the named objects.

Comment (evshi) [reply packages/runtime/src/widgets/Form/Radio.tsx:113]: @sagardspeed2 let's extend useTemplateData to handle an optional `value` prop from itemTemplate so it evaluates for you.

File: packages/framework/src/hooks/useTemplateData.ts [hunk 5]
```diff
@@ -46,13 +48,25 @@ export const useTemplateData = ({
...
   );
+
   const evaluatedNamedData = useEvaluate({ namedData });
 
   return {
     rawData,
-    namedData: evaluatedNamedData.namedData,
+    namedData: map(evaluatedNamedData.namedData, (val) => ({
+      ...val,
+      value: value
```
Comment (evshi) [inline packages/framework/src/hooks/useTemplateData.ts:61]: Pass this into useEvaluate. Make the key `_ensembleValue` to avoid conflicts with the evaluated keys.

**Review by evshi — CHANGES_REQUESTED** [review]

File: packages/framework/src/hooks/useTemplateData.ts (hunk 5 above)
Comment (sagardspeed2) [reply packages/framework/src/hooks/useTemplateData.ts:61]: @evshi that will not work. because of, as per your suggestion,

```js
  const namedData = useMemo(
    () =>
      map(rawData, (val: unknown) => ({ [name]: val, _ensembleValue: value })),
    [name, rawData],
  );

  const evaluatedNamedData = useEvaluate({ namedData });
```

when I will pass `_ensembleValue` in useEvaluate like this way, where yaml is like,

```yaml
    Radio:
      id: radioGroup
      label: Radio Group With Template
      item-template:
        data: getData.body.users.slice(0, 4)
        name: user
        value: ${user.email}
        template:
          Row:
            children:
              - Image:
                  width: 25
                  height: 25
                  source: ${user.image}
              - Text:
                  text: ${user.firstName} ${user.lastName}
                  
```

so each object of namedData will become like this before getting evaluated by `useEvaluate`

```js
{
user: {...},
_ensembleValue: ${user.email}
}
```

then when we pass this to useEvaluate, it will extract expressions like this,
image

and then this expression will convert into bindings and will be try to evaluate with ensemble storage, ensemble env, defaultScreenContext, ensemble user, ensemble secrets, device and so on.

and eventually it return the `_ensembleValue` as `null` because this value is depend on `user` of `namedData` which is not available in context.

**Comment by sagardspeed2** [comment]
@evshi I tried to update `useItemTemplateData` and `useEvalaute` hook to achieve your suggestion idea.
https://github.com/EnsembleUI/ensemble-react/pull/852

**Comment by evshi** [comment]
Please close this PR if #852 covers all the changes.

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 6]
```diff
@@ -38,19 +98,55 @@ export const RadioWidget: React.FC<RadioWidgetProps> = (props) => {
...
+  // extract template data
+  const { namedData } = useTemplateData({
+    data: itemTemplate?.data,
+    name: itemTemplate?.name,
+  });
 
-  useEffect(() => {
-    setValue(values?.initialValue?.toString());
-  }, [values?.initialValue]);
+  const RadioComponent = useMemo(() => withTemplate(Radio), []);
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:107]: Remove memo.

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 7]
```diff
@@ -38,19 +98,55 @@ export const RadioWidget: React.FC<RadioWidgetProps> = (props) => {
...
+      ...(styles?.visible === false ? { display: "none" } : undefined),
+    });
+
+    if (values?.items) {
+      values.items.forEach((item) => {
+        radioOptions.push({
+          item,
+          children: item.label,
+          disabled: values.enabled === false || item.enabled === false,
+          value: String(item.value),
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:124]: Do not coerce value to string.

File: packages/runtime/src/widgets/Form/Radio.tsx [hunk 8]
```diff
@@ -38,19 +98,55 @@ export const RadioWidget: React.FC<RadioWidgetProps> = (props) => {
...
       });
     }
-  }, [value, formInstance]);
+
+    if (isObject(itemTemplate) && !isEmpty(namedData)) {
+      map(namedData, (item: { [key: string]: unknown }) => {
+        const typedItem = get(item, itemTemplate.name) as CustomScope;
+        radioOptions.push({
+          item,
+          disabled: values?.enabled === false || typedItem.enabled === false,
```
Comment (evshi) [inline packages/runtime/src/widgets/Form/Radio.tsx:135]: Don't check `typedItem` for enabled - it's data it can't be guaranteed to have enabled prop.

**Review by evshi — CHANGES_REQUESTED** [review]

**Review by evshi — APPROVED** [review]
